theseus/optimizer/nonlinear/trust_region.py

```diff
@@ -0,0 +1,147 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+#
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+import abc
+from typing import Any, Dict, Optional, Type
+
+import torch
+
+from theseus.core import Objective
+from theseus.optimizer import Linearization
+from theseus.optimizer.linear import LinearSolver
+
+from .nonlinear_least_squares import NonlinearLeastSquares
+
+
+# See Nocedal and Wright, Numerical Optimization, Chapter 4
+# https://www.csie.ntu.edu.tw/~r97002/temp/num_optimization.pdf
+
+# This optimizer optimize() receives the following keywords (see p. 69):
+#       - trust_region_init (default 1.0)
+#       - accept_threshold (default 0.25)
+#       - shrink_threshold (default 0.25)
+#       - expand_threshold (default 0.75)
+#       - shrink_ratio (default 0.25)
+#       - expand_ratio (default 2.0)
+#
+# Iff the ratio between the actual error reduction after the step and the
+# predicted one is greater than `accept_threshold`, the step is applied.
+# Iff it's also greater than `expand_threshold`, the trust region is also
+# increased by `expand_ratio`.
+# Iff the ratio is lower than `shrink_threshold`,
+# then the trust region is reduced by `shrink_ratio`.
+# The trust region is initialized to `trust_region_init`.
+class TrustRegion(NonlinearLeastSquares, abc.ABC):
+    _MIN_TRUST_REGION = 1.0e-5
+    _MAX_TRUST_REGION = 1.0e5
+
+    def __init__(
+        self,
+        objective: Objective,
+        linear_solver_cls: Optional[Type[LinearSolver]] = None,
+        vectorize: bool = False,
+        linearization_cls: Optional[Type[Linearization]] = None,
+        linearization_kwargs: Optional[Dict[str, Any]] = None,
+        linear_solver_kwargs: Optional[Dict[str, Any]] = None,
+        abs_err_tolerance: float = 1e-10,
+        rel_err_tolerance: float = 1e-8,
+        max_iterations: int = 20,
+        step_size: float = 1.0,
```

Comment:
Add `self.max_trust_region` and `self.min_trust_region` here.

Reply:
Added.

Reply:
I'm still a little confused. Where the `self.max_trust_region` and `self.min_trust_region`. Also I wonder if we need `self.max_trust_region` since it does not matter if the `trust_reigon` is very large as long as the linearized model is correct.